### DISCUSSION TITLE
How to avoid an authentication middleware to ask credentials on swagger ui route

### DISCUSSION THREAD

**Discussion by SamaritanK** [fastapi/fastapi#9827] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
app = FastAPI()

app.add_middleware(
    OAuth2Middleware,
    public_key=AuthService().get_public_key()
)

# example 1 => middleware
@app.get('/test_1')
def test():
    return {'message': 'it works'}

# example 2 => Depends
@app.get('/test_2')
def test(user = Depends(__auth_service.get_current_user)):
    return {'message': 'it works'}


# example 3 => sub router
app.include_router(
    secured.router,
    dependencies=[Depends(AuthService().get_current_user)]
)

# example 4 => static folder
app.mount("/static", StaticFiles(directory="static"), name="static")
```

### Description

I'm building an API with auth based on OAuth2 JWT tokens.
To protect my routes the most elegant solution seems to me using middlewares (the alternative should be using Depends on each route), this is also because with a middleware it's possible to add request.user and request.auth and  I can add @requires to check scopes (similar to Spring security)

```
@requires('test_scope')
@app.get('/test_1')
def test():
    return {'message': 'it works'}
```

But adding this middleware is also requiring a token for the Swagger ui  (/doc) and any static mounted folder (example 4).
This seems to me quite of a standard situation (API under auth + public routes). 

There are all the tests I made after reading docs, discussions and stackoverflow.
https://github.com/tiangolo/fastapi/discussions/7691
https://fastapi.tiangolo.com/advanced/security/
https://fastapi.tiangolo.com/tutorial/middleware/

**Multiple app(s)**
pros:
- solves the static folder
- solves public routes
- enables @requires

cons:
- swagger ui for sub app with auth is still not public
- when mounted on the main app the route url changes 

**Route Depends parameter (example 2)**   
pros
- solves the static folder
- solves public routes
- solves swagger ui not public

cons
- doesn't enable @requires
- adds an unused parameter 

**APIRouter mount with dependencies (example 3)**
pros
- solves the static folder
- solves public routes
- solves swagger ui not public
- no unused parameter

cons
- doesn't enable @requires
- user is not available in the route (i.e. as an authorization parameter)  

### Operating System

macOS

### FastAPI Version

0.98.0

### Python Version

3.10.9

**Comment by SamaritanK** [comment]
I made a small change to example 3 that works better but it'd be very useful to hear other opinions because there's a lot around about security. 

main.py
```
def check_authentication(self, request: Request, token: HTTPAuthorizationCredentials = Security(HTTPBearer())):
        user = get_current_user(token) # function to decode and validate token
        request.scope["user"] = user # token added to request.scope (similar to an AuthenticationMiddleware)


app.include_router(
    secured.router,
    dependencies=[Depends(check_authentication)]
)
```

secured.py
```
@router.get(
    "/users/me",
)
def get_current_user(request: Request, user = Depends(get_current_user)):
    return {
        'depends': user,
        'request': request.user
    }
```

**Comment by skarfie123** [comment]
I like the idea of examples 2 or 3. depending on whether you think individual routes will have different scopes requirements or can be grouped into routers.

As for specfiying the scopes, you could use a callable class
```python
class RequiredScopes:
    def __init__(self, scopes: list[str]):
        self.scopes = scopes

    def __call__(self, user = Depends(get_current_user)) -> bool:
        # check the scopes
        pass
```
then on the router/route you can have `Depends(RequiredScopes(["foo", "bar"]))`

**Reply by SamaritanK** [reply]
I'd prefer something like

```
@router.get(
    "/helloWorld",
)
@Requires(['foo', 'bar'])
def hello_world():
    return "HELLO"
```

**Comment by YuriiMotov** [comment, marked as answer]
I think dependencies along with [`Security`](https://fastapi.tiangolo.com/reference/dependencies/#security), ability to specify [dependencies for router ](https://fastapi.tiangolo.com/tutorial/bigger-applications/#dependencies) and [dependency caching mechanism](https://fastapi.tiangolo.com/tutorial/dependencies/sub-dependencies/#using-the-same-dependency-multiple-times), is more than enough to solve this.
